File: infrastructure/monitoring/test-app/common/telemetry.py
import logging
import os

from opentelemetry import metrics, trace
from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.logging import LoggingInstrumentor
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor

try:
    from opentelemetry.exporter.otlp.proto.http._log_exporter import OTLPLogExporter
except ImportError:  # newer package layouts dropped the leading underscore
    from opentelemetry.exporter.otlp.proto.http.log_exporter import OTLPLogExporter


def _parse_headers(raw: str) -> dict:
    headers = {}
    for pair in raw.split(","):
        if not pair.strip():
            continue
        key, _, value = pair.partition("=")
        headers[key.strip()] = value.strip()
    return headers


def setup_telemetry(service_name: str):
    """Wires up OTLP traces, metrics and logs for one service, plus stdout logging."""
    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT", "http://alloy:4318").rstrip("/")
    headers = _parse_headers(os.environ.get("OTEL_EXPORTER_OTLP_HEADERS", ""))

    resource = Resource.create(
        {
            "service.name": service_name,
            "service.namespace": "shop-demo",
            "deployment.environment": os.environ.get("DEPLOYMENT_ENV", "demo"),
        }
    )

    logging.basicConfig(
        level=logging.INFO,
        format=f"%(asctime)s %(levelname)s [{service_name}] %(message)s",
    )

    # Traces
    tracer_provider = TracerProvider(resource=resource)
    tracer_provider.add_span_processor(
        BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{endpoint}/v1/traces", headers=headers))
    )
    trace.set_tracer_provider(tracer_provider)

    # Metrics
    reader = PeriodicExportingMetricReader(
        OTLPMetricExporter(endpoint=f"{endpoint}/v1/metrics", headers=headers),
        export_interval_millis=5000,
    )
    metrics.set_meter_provider(MeterProvider(resource=resource, metric_readers=[reader]))

    # Logs - forwarded to Loki (via Alloy) in addition to the stdout handler above
    logger_provider = LoggerProvider(resource=resource)
    logger_provider.add_log_record_processor(
        BatchLogRecordProcessor(OTLPLogExporter(endpoint=f"{endpoint}/v1/logs", headers=headers))
    )
    otlp_handler = LoggingHandler(level=logging.INFO, logger_provider=logger_provider)
    logging.getLogger().addHandler(otlp_handler)

    # Injects trace_id/span_id into stdlib log records so they show up in the
    # console handler too, and so Loki logs can be correlated to Tempo traces.
    LoggingInstrumentor().instrument(set_logging_format=True)

    return trace.get_tracer(service_name), metrics.get_meter(service_name)

File: peninei/backend/src/telemetry.ts
// src/telemetry.ts
// Wires up OTLP metrics and logs (no traces) shipped to the shared
// infrastructure/monitoring stack's Alloy collector. Must be imported
// first, before express/http, so the auto-instrumentations can patch
// those modules before they're used.
import { diag, DiagConsoleLogger, DiagLogLevel, metrics } from "@opentelemetry/api";
import { logs } from "@opentelemetry/api-logs";
import { resourceFromAttributes } from "@opentelemetry/resources";
import { MeterProvider, PeriodicExportingMetricReader } from "@opentelemetry/sdk-metrics";
import { LoggerProvider, BatchLogRecordProcessor } from "@opentelemetry/sdk-logs";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-http";
import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-http";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { HttpInstrumentation } from "@opentelemetry/instrumentation-http";
import { ExpressInstrumentation } from "@opentelemetry/instrumentation-express";
import { WinstonInstrumentation } from "@opentelemetry/instrumentation-winston";

const SERVICE_NAME = "peninei-backend";

const endpoint = (
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT || "http://alloy:4318"
).replace(/\/$/, "");

function parseHeaders(raw: string | undefined): Record<string, string> {
    const headers: Record<string, string> = {};
    (raw || "").split(",").forEach((pair) => {
        if (!pair.trim()) return;
        const [key, ...rest] = pair.split("=");
        headers[key.trim()] = rest.join("=").trim();
    });
    return headers;
}

const headers = parseHeaders(process.env.OTEL_EXPORTER_OTLP_HEADERS);

if (process.env.BACKEND_OTEL_DEBUG === "true") {
    diag.setLogger(new DiagConsoleLogger(), DiagLogLevel.INFO);
}

const resource = resourceFromAttributes({
    "service.name": SERVICE_NAME,
    "service.namespace": "peninei",
    "deployment.environment": process.env.DEPLOYMENT_ENV || "production",
});

// Metrics
const metricReader = new PeriodicExportingMetricReader({
    exporter: new OTLPMetricExporter({
        url: `${endpoint}/v1/metrics`,
        headers,
    }),
    exportIntervalMillis: 5000,
});
const meterProvider = new MeterProvider({ resource, readers: [metricReader] });
metrics.setGlobalMeterProvider(meterProvider);

// Logs - Winston log records are forwarded here in addition to the
// existing console/file transports (see logger.js), via instrumentation-winston.
const loggerProvider = new LoggerProvider({
    resource,
    processors: [
        new BatchLogRecordProcessor({
            exporter: new OTLPLogExporter({
                url: `${endpoint}/v1/logs`,
                headers,
            }),
        }),
    ],
});
logs.setGlobalLoggerProvider(loggerProvider);

registerInstrumentations({
    instrumentations: [
        new HttpInstrumentation(),
        new ExpressInstrumentation(),
        new WinstonInstrumentation(),
    ],
});

export const meter = metrics.getMeter(SERVICE_NAME);

// HTTP request metrics (drive req/s and latency dashboards in Grafana)
export const httpRequestsTotal = meter.createCounter("http_requests_total", {
    description: "Total HTTP requests handled",
});
export const httpRequestDuration = meter.createHistogram(
    "http_request_duration_ms",
    { description: "HTTP request duration", unit: "ms" }
);

// Halacha sync endpoint
export const syncRequestsTotal = meter.createCounter("sync_requests_total", {
    description: "Total /halachot/sync requests",
});
export const syncDuration = meter.createHistogram("sync_duration_ms", {
    description: "Duration of /halachot/sync requests",
    unit: "ms",
});

// Scraper job
export const halachasScrapedTotal = meter.createCounter(
    "halachas_scraped_total",
    { description: "Total halachot saved by the scraper" }
);
export const scrapeDuration = meter.createHistogram(
    "halacha_scrape_duration_ms",
    { description: "Duration of a per-date scrape", unit: "ms" }
);

// AI parsing job
export const halachasAiParsedTotal = meter.createCounter(
    "halachas_ai_parsed_total",
    { description: "Total halachot processed by the AI parser" }
);
